docs/programmers_guide/witness_formal_spec.md

```diff
@@ -106,63 +201,103 @@ Helper functions MUST have at least one argument.
 Helper functions MAY have variadic parameters: `HELPER_EXAMPLE(arg1, arg2, list...)`.
 Helper functions MAY contain recursion.
 
-## Data types
+## Instructions
 
-INTEGER - we treat the domain of integers as infinite, the overflow behaviour or mapping to the actual data types is undefined in this spec and should be dependent on implementation.
+### `LEAF key raw_value`
 
-## Execution flow 
+**Substitution rules**
 
-Let's look at the example.
+```
+LEAF(key, raw_value) |=> LeafNode{key, ValueNode(raw_value)}
+```
 
-Our example witness would be `HASH h1; HASH h2; BRANCH 0b11`.
+### `EXTENSION key`
 
-Initial state of the stack is ` <empty> `;
+**Substitution rules**
 
----
+```
+GUARD node != nil
 
-**1. Executing `HASH h1`**: push a `hashNode` to the stack.
+Node(node) EXTENSION(key) |=> ExtensionNode{key, node}
 
-Witness: `HASH h2; BRANCH 0b11`
+```
 
-Stack: `(hashNode(h1), h1)`
+### `HASH raw_hash`
 
----
+Pushes a `HashNode` to stack.
 
-**2. Executing `HASH h2`**: push one more `hashNode` to the stack.
+**Substitution rules**
 
-Witness `BRANCH 0b11`
+```
+HASH(hash_value) |=> HashNode{hash_value}
+```
+
+### `CODE raw_code`
+
+Pushes an nil node + the code hash to the stack.
```

Comment:
Is this correct? It pushes CodeNode on the stack (and there isn't stack anymore)